### PR TITLE
Resurrect parallel distributed insert select with s3Cluster

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Originally merged here #39107 but reverted in #40577

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Flaky check 😖

File: src/Storages/IStorageCluster.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace DB
+{
+
+
+/**
+ *  Base cluster for Storages used in table functions like s3Cluster and hdfsCluster
+ *  Needed for code simplification around parallel_distributed_insert_select
+ */
+class IStorageCluster: public IStorage
```
Comment (antonio2368) [inline src/Storages/IStorageCluster.h:15]:
```suggestion
class IStorageCluster : public IStorage
```

File: src/Storages/HDFS/StorageHDFSCluster.cpp [hunk 2]
```diff
@@ -140,6 +134,29 @@ QueryProcessingStage::Enum StorageHDFSCluster::getQueryProcessingStage(
 }
 
 
+void StorageHDFSCluster::createIteratorAndCallback(ContextPtr context) const
+{
+    cluster = context->getCluster(cluster_name)->getClusterWithReplicasAsShards(context->getSettingsRef());
+
+    iterator = std::make_shared<HDFSSource::DisclosedGlobIterator>(context, uri);
```
Comment (antonio2368) [inline src/Storages/HDFS/StorageHDFSCluster.cpp:141]: do you need to have iterator as member? you just copy it into the callback that's all the usage I see.

File: src/Storages/HDFS/StorageHDFSCluster.cpp [hunk 3]
```diff
@@ -140,6 +134,29 @@ QueryProcessingStage::Enum StorageHDFSCluster::getQueryProcessingStage(
 }
 
 
+void StorageHDFSCluster::createIteratorAndCallback(ContextPtr context) const
```
Comment (antonio2368) [inline src/Storages/HDFS/StorageHDFSCluster.cpp:137]: Why can't this function return cluster and callback? It's a bit confusing to have it modify member values while calling it always before using them.
Don't see a lifetime issue as the `shared_ptr` are copied/used instantly.

Or even better, 2 functions for iterator and cluster.

File: src/Storages/StorageDistributed.cpp [hunk 4]
```diff
@@ -882,6 +865,114 @@ std::optional<QueryPipeline> StorageDistributed::distributedWrite(const ASTInser
...
+    }
+    else if (auto src_storage_cluster = std::dynamic_pointer_cast<IStorageCluster>(src_storage))
+    {
+        return distributedWriteFromClusterStorage(*src_storage_cluster, query, local_context);
+    }
+    else if (local_context->getClientInfo().distributed_depth == 0)
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Parallel distributed INSERT SELECT is not possible. "\
+            "Reason: distributed reading is supported only from Distributed engine or *Cluster table functions, but got {} storage", src_storage->getName());
+    }
```
Comment (antonio2368) [inline src/Storages/StorageDistributed.cpp:970]: No need for else

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 5]
```diff
@@ -4450,6 +4455,106 @@ SinkToStoragePtr StorageReplicatedMergeTree::write(const ASTPtr & /*query*/, con
...
+    if (!src_storage)
+        return {};
+
+    if (auto src_distributed = std::dynamic_pointer_cast<IStorageCluster>(src_storage))
+    {
+        return distributedWriteFromClusterStorage(src_distributed, query, local_context);
+    }
+    else if (local_context->getClientInfo().distributed_depth == 0)
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Parallel distributed INSERT SELECT is not possible. Reason: distributed "\
```
Comment (antonio2368) [inline src/Storages/StorageReplicatedMergeTree.cpp:4550]:
```suggestion
        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Parallel distributed INSERT SELECT is not possible. Reason: distributed "
```

File: src/Storages/StorageS3Cluster.cpp [hunk 6]
```diff
@@ -170,6 +165,29 @@ QueryProcessingStage::Enum StorageS3Cluster::getQueryProcessingStage(
 }
 
 
+void StorageS3Cluster::createIteratorAndCallback(ASTPtr query, ContextPtr context) const
```
Comment (antonio2368) [inline src/Storages/StorageS3Cluster.cpp:168]: same comment as for HDFS, API is a bit confusing for me.

File: src/Storages/HDFS/StorageHDFSCluster.cpp (hunk 3 above)
Comment (nikitamikhaylov) [reply src/Storages/HDFS/StorageHDFSCluster.cpp]: I was a mess when I wrote this. Generally I thought this "initialization" could happen twice and thus made such an API

**Comment by nikitamikhaylov** [comment]
Stateful test (asan) - https://github.com/ClickHouse/ClickHouse/issues/41500
Stress test (ubsan) - https://github.com/ClickHouse/ClickHouse/issues/42009

**Comment by nikitamikhaylov** [comment]
@antonio2368 This is PR is a second stage after revert directed only to fix flaky tests. Merging without review...

**Comment by tavplubix** [comment]
It breaks `test_s3_cluster/test.py::test_distributed_insert_select`:
https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZEYXkoY2hlY2tfc3RhcnRfdGltZSkgYXMgZCwKY291bnQoKSwgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlciksICBhbnkocmVwb3J0X3VybCkKZnJvbSBjaGVja3Mgd2hlcmUgJzIwMjItMDYtMDEnIDw9IGNoZWNrX3N0YXJ0X3RpbWUgYW5kIHRlc3RfbmFtZSBsaWtlICcldGVzdF9zM19jbHVzdGVyL3Rlc3QucHk6OnRlc3RfZGlzdHJpYnV0ZWRfaW5zZXJ0X3NlbGVjdCUnIGFuZCB0ZXN0X3N0YXR1cyBpbiAoJ0ZBSUwnLCAnRkxBS1knKSBncm91cCBieSBkIG9yZGVyIGJ5IGQgZGVzYw==

**Pull request merged**